fastlane_bot/modes/pairwise_multi_bal.py

```diff
@@ -0,0 +1,217 @@
+        combos = self.get_balancer_combos(self.CCm, self.flashloan_tokens)
+
+        #print(f"bal combos = {len(combos), combos}")
+        candidates = []
+        self.ConfigObj.logger.debug(
+            f"\n ************ combos: {len(combos)} ************\n"
+        )
+        for tkn0, tkn1 in combos:
+            r = None
+            CC = self.CCm.bypairs(f"{tkn0}/{tkn1}")
+            if len(CC) < 2:
+                continue
+            carbon_curves = [x for x in CC.curves if x.params.exchange == "carbon_v1"]
+            not_bal_curves = [
+                x for x in CC.curves if x.params.exchange != "balancer"
+            ]
+
+            bal_curves = [x for x in CC.curves if x.params.exchange == "balancer"]
+            curve_combos = [[bal_curve] + [curve] for curve in not_bal_curves for bal_curve in bal_curves]
+            curve_combos += [[bal_curve] + carbon_curves for bal_curve in bal_curves]
+            #print(bal_curves)
+
+            for curve_combo in curve_combos:
+                src_token = tkn1
+                if len(curve_combo) < 2:
+                    continue
+                #print(f"curve combo = {curve_combo}")
+
+                try:
+                    (
+                        O,
+                        profit_src,
+                        r,
+                        trade_instructions_df,
+                    ) = self.run_main_flow(curves=curve_combo, src_token=src_token, tkn0=tkn0, tkn1=tkn1)
+
+                    # if profit_src > self.ConfigObj.DEFAULT_MIN_PROFIT:
+                    #
+                    #     print(profit_src, r)
+
+                    non_carbon_cids = [
+                        curve.cid
+                        for curve in curve_combo
+                        if curve.params.get("exchange") != "carbon_v1"
+                    ]
+                    non_carbon_row = trade_instructions_df.loc[non_carbon_cids[0]]
+                    tkn0_into_carbon = non_carbon_row[0] < 0
+                    wrong_direction_cids = self.get_wrong_direction_cids(
+                        tkn0_into_carbon, trade_instructions_df
+                    )
+
+                    if non_carbon_cids and len(wrong_direction_cids) > 0:
+                        filtered_curves = self.process_wrong_direction_pools(
+                            curve_combo=curve_combo, wrong_direction_cids=wrong_direction_cids
+                        )
+                        if len(filtered_curves) < 2:
+                            continue
+                        (
+                            O,
+                            profit_src,
+                            r,
+                            trade_instructions_df,
+                        ) = self.run_main_flow(curves=filtered_curves, src_token=src_token, tkn0=tkn0, tkn1=tkn1)
+
+                    trade_instructions_dic = r.trade_instructions(O.TIF_DICTS)
+                    trade_instructions = r.trade_instructions()
+
+                except Exception:
```

Comment:
should be a specific type of Exception that we expect, and even then it should log something in debug at least. probably this needs to be changed in other modes too, but might as well get it changed from the start in this mode...